packages/chain-mon/src/initialized-upgraded-mon/service.ts

```diff
@@ -0,0 +1,230 @@
+          labels: ['label', 'address'],
+        },
+        upgradedCalls: {
+          type: Gauge,
+          desc: 'Successful transactions to tracked contracts emitting upgraded event',
+          labels: ['label', 'address'],
+        },
+        unexpectedRpcErrors: {
+          type: Counter,
+          desc: 'Number of unexpected RPC errors',
+          labels: ['section', 'name'],
+        },
+      },
+    })
+  }
+
+  protected async init(): Promise<void> {
+    // Connect to L1.
+    await waitForProvider(this.options.rpc, {
+      logger: this.logger,
+      name: 'L1',
+    })
+
+    this.state.chainId = await getChainId(this.options.rpc)
+
+    const l1StartingBlockTag = networks[this.state.chainId].l1StartingBlockTag
+
+    if (this.options.startBlockNumber === -1) {
+      const block_number =
+        l1StartingBlockTag != null
+          ? (await this.options.rpc.getBlock(l1StartingBlockTag)).number
+          : 0
+      this.state.highestUncheckedBlockNumber = block_number
+    } else {
+      this.state.highestUncheckedBlockNumber = this.options.startBlockNumber
+    }
+
+    this.state.contracts = JSON.parse(this.options.contracts)
+  }
+
+  protected async main(): Promise<void> {
+    if (
+      (await this.options.rpc.getBlockNumber()) <
+      this.state.highestUncheckedBlockNumber
+    ) {
+      this.logger.info('Waiting for new blocks')
+      return
+    }
+
+    const block = await this.options.rpc.getBlock(
+      this.state.highestUncheckedBlockNumber
+    )
+    this.logger.info('Checking block', {
+      number: block.number,
+    })
+
+    const transactions: TransactionResponse[] = []
+    for (const txHash of block.transactions) {
+      const t = await this.options.rpc.getTransaction(txHash)
+      transactions.push(t)
+    }
+
+    for (const transaction of transactions) {
+      for (const contract of this.state.contracts) {
+        const to =
+          transaction.to != null ? transaction.to : transaction['creates']
+        if (compareAddrs(contract.address, to)) {
+          try {
+            const transactionReceipt = await transaction.wait()
+            for (const log of transactionReceipt.logs) {
+              if (log.topics.includes(topic_initialized)) {
+                this.metrics.initializedCalls.inc({
+                  label: contract.label,
+                  address: contract.address,
+                })
+                this.logger.info('initialized event', {
+                  label: contract.label,
+                  address: contract.address,
+                })
+              } else if (log.topics.includes(topic_upgraded)) {
+                this.metrics.upgradedCalls.inc({
+                  label: contract.label,
+                  address: contract.address,
+                })
+                this.logger.info('upgraded event', {
+                  label: contract.label,
+                  address: contract.address,
+                })
+              }
```

Comment:
Curious what the advantage is of iterating over all the block logs rather than using an event filter of some kind is?